### ISSUE TITLE
Possible parallel processing?

### ISSUE THREAD

**Issue by fmmattioni** [easystats/correlation#84]
Hi, I really love the package! Thanks for that!

I have noticed that when I have a large data frame (many columns) the `correlation::correlation()` function takes a while to process. I assume it is not the goal of the package to include some sort of parallel processing but I was wondering if anyone knows if I could use `furrr`, for example, to speed up the code?

**Comment by DominiqueMakowski** [comment]
That's a good question, and unfortunately totally outside my skills. Maybe @strengejacke or @mattansb would know a bit more?

**Comment by fmmattioni** [comment]
Hi @mattansb, thanks for the reply.

The data frame I am working on has approximately 400 columns and 50 rows. The data frame is grouped by a column that has two levels, so 25 rows for each group. I am performing a simple Pearson correlation now.

Perhaps I could parallel process the two groups?

**Comment by mattansb** [comment]
@DominiqueMakowski there is significant time diffs in performance.... Any idea what is happening here?

``` r
M <- data.frame(matrix(rnorm(400), ncol = 40))

print(t <- Sys.time())
#> [1] "2020-06-08 13:23:16 IDT"
rr <- correlation::correlation(M)
print(t <- Sys.time() - t)
#> Time difference of 10.3551 secs


print(t <- Sys.time())
#> [1] "2020-06-08 13:23:27 IDT"
rr <- psych::corr.test(M)
print(t <- Sys.time() - t)
#> Time difference of 0.6146462 secs
```

<sup>Created on 2020-06-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by strengejacke** [comment]
The bottleneck here was definitely the call to `parameters::model_parameters()`. I re-implemented just the code for cleaning correlations, and speed-increase is quite good!

```r
M <- data.frame(matrix(rnorm(400), ncol = 40))
microbenchmark::microbenchmark(
  easystats = correlation::correlation(M),
  psych = psych::corr.test(M),
  times = 100
)
#> Unit: milliseconds
#>       expr      min       lq      mean   median        uq      max neval cld
#>  easystats 890.8372 911.2753 932.10221 917.7569 938.23890 1246.480   100   b
#>      psych  30.5107  31.5382  35.19228  32.0385  33.85325  167.962   100  a 
```

Mean computation time on my system now is about 0.93 seconds for `correlation::correlation()` (coming from ~ 2.5 seconds).

The Profile-feature in RStudio is really great for finding such bottlenecks. While in general we do not optimize for speed, and I think regarding the scope of our packages, we mostly don't need to do so, there might be some functions where we may try to improve performance, in the long run...

**Comment by strengejacke** [comment]
Here a comparison between the "old" implementation, and the new one:

``` r
M <- data.frame(matrix(rnorm(400), ncol = 40))
microbenchmark::microbenchmark(
  easystats_old = correlation::correlation(M, do_profile = TRUE),
  easystats_new = correlation::correlation(M),
  psych = psych::corr.test(M),
  times = 20
)
#> Unit: milliseconds
#>           expr       min         lq       mean    median        uq       max
#>  easystats_old 2302.0921 2321.15905 2355.80649 2331.0354 2383.6593 2538.1921
#>  easystats_new  925.1745  938.61810  948.87694  941.7087  961.0896  981.1898
#>          psych   31.8199   33.66735   39.05076   34.4887   37.8351  112.5483
```

<sup>Created on 2020-11-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by strengejacke** [comment]
btw, whenever you don't use regex-patterns in `grepl()`, `gsub()`, `strplit()` etc., use `fixed = TRUE`, which will increase performance remarkably when called multiple times:

``` r
x <- c("abc", "def", "bcd")
microbenchmark::microbenchmark(
  fixed = grepl("b", "x", x, fixed = TRUE),
  non_fixed = grepl("b", "x", x),
  times = 1000
)
#> Unit: microseconds
#>       expr min  lq   mean median  uq  max neval cld
#>      fixed 1.0 1.1 1.1617    1.1 1.2  2.8  1000  a 
#>  non_fixed 2.6 2.7 2.8534    2.7 2.8 80.4  1000   b
```

<sup>Created on 2020-11-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by IndrajeetPatil** [comment]
I think once the entire ecosystem is stable (all > `1.0.0`), we can revisit performance issues in the releases thereafter and clean up possible bottlenecks like these. 

At that time, it might also be worth it to have some `C++` expertise on board to advise us on this topic.
